cpp/src/listener.cpp

```diff
@@ -52,7 +45,31 @@ Listener::Listener(std::shared_ptr<Worker> worker,
   setParent(worker);
 }
 
-Listener::~Listener() { ucxx_debug("Listener destroyed: %p", _handle.get()); }
+Listener::~Listener()
+{
+  decltype(_handle) origHandle = _handle;
+
+  auto worker = std::static_pointer_cast<Worker>(_parent);
+
+  if (worker->isProgressThreadRunning()) {
+    worker->registerGenericPre([this]() {
+      ucp_listener_destroy(this->_handle);
+      this->_handle = nullptr;
+    });
+    while (_handle != nullptr)
```

Comment:
And here. This is likely UB anyway since `_handle` is not volatile: I suspect the compiler is free to assume that `_handle == nullptr` since infinite loops are UB.